### PR TITLE
GSdx-d3d11: Some HLSL shader work

### PR DESCRIPTION
**Pull request by KrossX**
Make the D3D shader closer to OpenGL's one, possibly fixing some bugs during previous shader port. Also, quite possibly adding new bugs. Yay!

Improves bloom/blur in DBZ BT2, SOTC matching OpenGL visuals.
Improves https://github.com/PCSX2/pcsx2/issues/2864 matching OpenGL visuals.

### CODE REVIEW

File: plugins/GSdx/res/tfx.fx [hunk 1]
```diff
@@ -632,37 +625,145 @@ float4 ps_color(PS_INPUT input)
...
-	if(PS_CLR1) // needed for Cd * (As/Ad/F + 1) blending modes
+void ps_fbmask(inout float4 C, float2 pos_xy)
+{
+	if (PS_FBMASK)
 	{
-		c.rgb = 1;
+		float4 rt = RtSampler.Load(int3(pos_xy, 0));
+		uint4 denorm_rt = uint4(rt * 255.0f + 0.1f);
+		uint4 denorm_c = uint4(C);
+		C = float4((denorm_c & ~FbMask) | (denorm_rt & FbMask)) / 255.0f;
```
Comment (lightningterror) [inline plugins/GSdx/res/tfx.fx]: / 255.0f not needed?

Comment (lightningterror) [reply plugins/GSdx/res/tfx.fx]: Actually you can use gl code here too for the entire function.

File: plugins/GSdx/res/tfx.fx [hunk 2]
```diff
@@ -632,37 +625,145 @@ float4 ps_color(PS_INPUT input)
...
+	float4 T = fetch_rgb(int2(input.p.xy));
 #elif PS_CHANNEL_FETCH == 6
-	float4 t = fetch_gXbY(int2(input.p.xy));
+	float4 T = fetch_gXbY(int2(input.p.xy));
 #elif PS_DEPTH_FMT > 0
-	float4 t = sample_depth(st_int, input.p.xy);
+	float4 T = sample_depth(st_int, input.p.xy);
 #else
-	float4 t = sample_color(st);
+#define CHUCHU
```
Comment (lightningterror) [inline plugins/GSdx/res/tfx.fx]: ???

File: plugins/GSdx/res/tfx.fx [hunk 3]
```diff
@@ -632,37 +625,145 @@ float4 ps_color(PS_INPUT input)
...
+	float4 T = sample_color(st);
 #endif
 
-	float4 c = tfx(t, input.c);
+	float4 xC = trunc(input.c * 255.0f + 0.5f); // FIX THIS
```
Comment (turtleli) [inline plugins/GSdx/res/tfx.fx]: If you're trying to follow the OpenGL/GLSL code, then in addition to changing the colour format to `DXGI_FORMAT_R8G8B8A8_UINT` the corresponding `VS_INPUT` entry must be changed to `uint4`.

Comment (KrossX) [reply plugins/GSdx/res/tfx.fx]: Nice! That did the trick.

File: plugins/GSdx/res/tfx.fx (hunk 2 above)
Comment (KrossX) [reply plugins/GSdx/res/tfx.fx]: Debug/testing leftover. I use very descriptive names.

File: plugins/GSdx/res/tfx.fx (hunk 1 above)
Comment (KrossX) [reply plugins/GSdx/res/tfx.fx]: Nope, not needed. Check the new code please.

Comment (lightningterror) [reply plugins/GSdx/res/tfx.fx]: Seems good now.

File: plugins/GSdx/res/tfx.fx [hunk 4]
```diff
@@ -52,11 +52,12 @@
 #endif
 
 #define SW_BLEND (PS_BLEND_A || PS_BLEND_B || PS_BLEND_D)
+#define PS_AEM_FMT (PS_FMT & 3)
 
 struct VS_INPUT
 {
 	float2 st : TEXCOORD0;
-	float4 c : COLOR0;
+	uint4 c : COLOR0;
```
Comment (lightningterror) [inline plugins/GSdx/res/tfx.fx]: Don't we need to change it across all shaders?

Comment (KrossX) [reply plugins/GSdx/res/tfx.fx]: I don't think so. The cpp code change is for `"tfx.fx", "vs_main"` so other shaders should be unaffected. I would prefer not to mess with those.

**Comment by lightningterror** [comment]
You can squash some of the commits, keep refactor commit separate and rename them.
Maybe:
tfx.fx: Refactor .... , 
GSdx-d3d11: Make shader code ....

**Comment by KrossX** [comment]
How about now?

**Comment by lightningterror** [comment]
Better but I prefer commit titles starting with `GSdx-d3d11:`

**Comment by KrossX** [comment]
And now? >_>

**Comment by lightningterror** [comment]
Looks good :+1:

File: plugins/GSdx/res/tfx.fx [hunk 5]
```diff
@@ -517,76 +518,61 @@ float4 sample_color(float2 st)
...
+		case 2:
+			if (a < AREF) discard;
+			break;
+		case 3:
+			if (abs(a - AREF) > 0.5f) discard;
+			break;
+		case 4:
+			if (abs(a - AREF) < 0.5f) discard;
+			break;
+	}
```
Comment (turtleli) [inline plugins/GSdx/res/tfx.fx]: I'd prefer if this formatting change was moved to the first commit.

Otherwise the commit looks ok.

**Comment by turtleli** [comment]
Truncation changes (the stuff from 36e83b42ce615e4a2a731e7028dd9cb4e83949a2) makes Suikoden Tactics (movement range fog) on D3D11 look similar to OpenGL.

File: plugins/GSdx/res/tfx.fx (hunk 5 above)
Comment (KrossX) [reply plugins/GSdx/res/tfx.fx]: I think it's done now. It required some git trickery I'm not comfortable with, so I hope no change got lost.

**Comment by lightningterror** [comment]
Here are some comparison improvements:

Silent Hill 4 shadows fixed (note: misses impossible blending for the darker image)

Master
gsdx_20190825203143

PR
gsdx_20190825203135

Opengl
gsdx_20190825203421


Dragonball Z Budokai Tenkaichi 2 (improves bloom/blur, matching opengl)
Master
gsdx_20190825202825

PR
gsdx_20190825202807


Shadow of the Colossus (improves bloom, needs sw blending to further improve the blooming like on opengl)

Master
gsdx_20190825201636

PR
gsdx_20190825201655

**Comment by KrossX** [comment]
Added missing `* 255.0f ` to `PS_URBAN_CHAOS_HLE`

**Comment by lightningterror** [comment]
I'll go ahead and merge this, PR has been tested extensively for a month and reviewed.
So far no issues have been spotted, improvements are visible across many games.

**Pull request merged**